File: uniffi_bindgen/src/bindings/python/templates/Slab.py
class UniffiSlab:
    MAX_ENTRIES = 1_000_000_000;
    MAX_SPECIAL_VALUES = 2**30 - MAX_ENTRIES;
    END_OF_LIST = 2**32 - 1
    TAG_MASK = 0xC0000000
    INDEX_MASK = 0x3FFFFFFF

    @dataclass
    class Vacant:
        next: int

    @dataclass
    class Occupied:
        value: object

    def __init__(self, tag=0):
        if tag >= 4:
            raise InternalError("Tag does not fit in 2 bits")
        self._tag = tag << 30
        self._next = 0
        self._entries = []
        self._lock = threading.Lock()

    def _get_index(self, handle: int) -> int:
        if handle & self.TAG_MASK == self._tag:
            return handle & self.INDEX_MASK
        else:
            raise InternalError("Invalid tag")

    def _get_handle(self, index: int) -> int:
        return index | self._tag

    def _get_vacant_to_use(self) -> int:
        if self._next == self.END_OF_LIST:
            self._entries.append(self.Vacant(self.END_OF_LIST))
            return len(self._entries) - 1
        else:
            old_next = self._next
            next_entry = self._entries[old_next]
            if not isinstance(next_entry, self.Vacant):
                raise InternalError("self._next is not Vacant")
            self._next = next_entry.next
            return old_next

    def insert(self, value: object) -> int:
        with self._lock:
            idx = self._get_vacant_to_use()
            self._entries[idx] = self.Occupied(value)
            return self._get_handle(idx)

    def remove(self, handle: int) -> object:
        idx = self._get_index(handle)
        with self._lock:
            entry = self._entries[idx]
            if not isinstance(entry, self.Occupied):
                raise InternalError("entry is Occupied, was the handle already removed?")
            self._entries[idx] = self.Vacant(self.next)
            self.next = idx
            return entry.value

    def get(self, handle: int) -> object:
        idx = self._get_index(handle)
        with self._lock:
            entry = self._entries[idx]
            if not isinstance(entry, self.Occupied):
                raise InternalError("entry is Occupied, was the handle already removed?")
            else:
                return entry.value

    def special_value(self, val: int) -> int:
        return self.MAX_ENTRIES + val

File: uniffi_bindgen/src/bindings/swift/templates/Slab.swift
let UNIFFI_SLAB_TAG_MASK: UInt32 = 0xC0000000
let UNIFFI_SLAB_INDEX_MASK: UInt32 = 0x3FFFFFFF
let UNIFFI_SLAB_MAX_ENTRIES: Int = 1_000_000_000
let UNIFFI_SLAB_END_OF_LIST: Int = Int.max

/// Store values and generate handles that can be used by the foreign code
struct UniffiSlab<T> {
    enum Entry {
        case vacant(next: Int)
        case occupied(value: T)
    }

    private let tag: UInt32
    private let lock = NSLock()
    private var next: Int = 0
    private var entries: [Entry] = []

    init() {
        tag = 0;
    }

    init(tag: UInt32) {
        if (tag >= 4) {
            fatalError("Tag can't fit in 2 bits")
        }
        self.tag = tag << 30;
    }

    private func getIndex(handle: UInt32) -> Int {
        if (handle & UNIFFI_SLAB_TAG_MASK == tag) {
            return Int(handle & UNIFFI_SLAB_INDEX_MASK)
        } else {
            fatalError("Invalid tag")
        }
    }

    private func getHandle(index: Int) -> UInt32 {
        return UInt32(index) | tag
    }

    private mutating func getVacantToUse() -> Int {
        if (next == UNIFFI_SLAB_END_OF_LIST) {
            entries.append(Entry.vacant(next: UNIFFI_SLAB_END_OF_LIST))
            return entries.count - 1
        } else {
            let oldNext = next
            switch entries[next] {
                case Entry.occupied: fatalError("next is occupied")
                case Entry.vacant(let next):
                    self.next = next
                    return oldNext
            }
        }
    }

    public mutating func insert(value: T) -> UInt32 {
        return lock.withLock {
            let idx = getVacantToUse()
            entries[idx] = Entry.occupied(value: value)
            return getHandle(index: idx)
        }
    }

    public mutating func remove(handle: UInt32) -> T {
        let idx = getIndex(handle: handle)
        return lock.withLock {
            switch entries[idx] {
                case Entry.vacant:
                    fatalError("Entry vacant (was the handle already removed?)")

                case Entry.occupied(let value):
                    entries[idx] = Entry.vacant(next: self.next)
                    self.next = idx
                    return value
            }
        }
    }

    public func get(handle: UInt32) -> T {
        let idx = getIndex(handle: handle)
        return lock.withLock {
            switch entries[idx] {
                case Entry.vacant:
                    fatalError("Entry vacant (was the handle already removed?)")

                case Entry.occupied(let value):
                    return value
            }
        }
    }

    public func specialValue(value: Int) -> UInt32 {
        return UInt32(UNIFFI_SLAB_MAX_ENTRIES + value)
    }
}
